packages/gatsby/src/bootstrap/load-plugins/validate.ts

```diff
@@ -165,6 +167,98 @@ export async function handleBadExports({
   }
 }
 
+const Joi: joi.Root = joi.extend({
+  // This tells Joi to extend _all_ types with .dotenv(), see
+  // https://github.com/sideway/joi/commit/03adf22eb1f06c47d1583617093edee3a96b3873
+  // @ts-ignore Joi types weren't updated with that commit, PR: https://github.com/sideway/joi/pull/2477
+  type: /^s/,
+  rules: {
+    dotenv: {
+      args: [`name`],
+      validate(value, helpers, args): void {
+        if (!args.name) {
+          return helpers.error(
+            `any.dotenv requires the environment variable name`
+          )
+        }
+        return value
+      },
+      method(name): Schema {
+        return this.$_addRule({ name: `dotenv`, args: { name } })
+      },
+    },
+  },
+})
+
+export async function validatePluginOptions({
+  flattenedPlugins,
+}: {
+  flattenedPlugins: Array<IPluginInfo & Partial<IFlattenedPlugin>>
+}): Promise<void> {
+  const errors = (
+    await Promise.all(
+      flattenedPlugins.map(
+        async (plugin): Promise<boolean | null> => {
+          if (
+            !plugin.nodeAPIs ||
+            plugin.nodeAPIs?.indexOf(`pluginOptionsSchema`) === -1
+          )
+            return null
+
+          const gatsbyNode = require(`${plugin.resolve}/gatsby-node`)
+          if (!gatsbyNode.pluginOptionsSchema) return null
+
+          const optionsSchema: joi.ObjectSchema = gatsbyNode.pluginOptionsSchema(
+            {
+              Joi,
+            }
+          )
+
+          // Validate correct usage of pluginOptionsSchema
+          if (!Joi.isSchema(optionsSchema) || optionsSchema.type !== `object`) {
+            reporter.warn(
+              `Plugin "${plugin.name}" has an invalid options schema so we cannot verify your configuration for it.`
+            )
+            return null
+          }
+
+          try {
+            // All plugins have plugins: [] added to their options by core, no need to validate it
+            const optionsSchemaWithPlugins = optionsSchema.append({
+              plugins: Joi.any(),
+            })
+
+            await validateOptionsSchema(
+              optionsSchemaWithPlugins,
+              plugin.pluginOptions
+            )
+          } catch (error) {
+            if (error instanceof Joi.ValidationError) {
```

Comment:
Does the structured error need implementing in cloud before this merges?

Reply:
Great question! Clarified internally that it does not 👍